### PR TITLE
Allow all api scripts to take dictionary

### PR DESCRIPTION
**Pull request by aloukina**
This PR addressed the following issues:

1. #356 is now extended to all main API scripts.

2. #363 is now applied to all parametrized tests.

3. Minor test refactoring to make things more streamlined.

### CODE REVIEW

File: rsmtool/test_utils.py [hunk 1]
```diff
@@ -372,8 +438,15 @@ def do_run_evaluation(source, experiment_id, config_file):
         Path to where the test is located on disk.
     experiment_id : str
         Experiment ID to use when running.
-    config_file : str
-        Path to the experiment configuration file.
+    config_input : str or Configuration or Dictionary
```
Comment (desilinguist) [inline rsmtool/test_utils.py]:
```suggestion
    config_input : str or Configuration or dict
```

File: rsmtool/test_utils.py [hunk 2]
```diff
@@ -395,8 +480,14 @@ def do_run_prediction(source, config_file):
     ----------
     source : str
         Path to where the test is located on disk.
-    config_file : str
-        Path to the experiment configuration file.
+    config_input : str or Configuration or Dictionary
```
Comment (desilinguist) [inline rsmtool/test_utils.py]:
```suggestion
    config_input : str or Configuration or dict
```

File: rsmtool/test_utils.py [hunk 3]
```diff
@@ -422,15 +524,34 @@ def do_run_comparison(source, config_file):
...
```
Comment (desilinguist) [inline rsmtool/test_utils.py]:
```suggestion
    config_input : str or Configuration or dict
```

File: rsmtool/test_utils.py [hunk 4]
```diff
@@ -439,8 +560,14 @@ def do_run_summary(source, config_file):
...
```
Comment (desilinguist) [inline rsmtool/test_utils.py]:
```suggestion
    config_input : str or Configuration or dict
```

File: tests/test_experiment_rsmcompare.py [hunk 5]
```diff
@@ -23,27 +28,105 @@
 _AUTO_UPDATE = False
 
 
+DIRS_TO_REMOVE = []
```
Comment (desilinguist) [inline tests/test_experiment_rsmcompare.py]: We don't need this right?

File: tests/test_experiment_rsmcompare.py [hunk 6]
```diff
@@ -23,27 +28,105 @@
...
+          'lr-subgroups-with-edge-cases_vs_lr-subgroups-with-edge-cases')
 ])
 def test_run_experiment_parameterized(*args, **kwargs):
     if TEST_DIR:
         kwargs['given_test_dir'] = TEST_DIR
     check_run_comparison(*args, **kwargs)
 
 
+def test_run_experiment_lr_compare_with_object():
+    # test rsmcompare using the Configuration object, rather than a file;
```
Comment (desilinguist) [inline tests/test_experiment_rsmcompare.py]: Let's covert this comment into a docstring?

File: tests/test_experiment_rsmeval.py [hunk 7]
```diff
@@ -20,6 +27,8 @@
     from rsmtool.test_utils import rsmtool_test_dir
 
 
+DIRS_TO_REMOVE = []
```
Comment (desilinguist) [inline tests/test_experiment_rsmeval.py]: Unneeded?

File: tests/test_experiment_rsmpredict.py [hunk 8]
```diff
@@ -96,6 +102,57 @@ def test_run_experiment_lr_rsmtool_and_rsmpredict():
...
+
+    # set up a temporary directory since
+    # we will be using getcwd
+    temp_dir = tempfile.TemporaryDirectory(prefix=getcwd())
+
+    old_file_dict = {'feature_file': 'data/files/test.csv',
+                     'experiment_dir': 'data/experiments/lr-predict-dict/existing_experiment'}
+
+    new_file_dict = copy_data_files(temp_dir.name,
+                                    old_file_dict)
```
Comment (desilinguist) [inline tests/test_experiment_rsmpredict.py]: You don't need a newline between the arguments.

File: tests/test_experiment_rsmsummarize.py [hunk 9]
```diff
@@ -44,42 +49,52 @@ def test_run_experiment_parameterized(*args, **kwargs):
...
+    # we will be using getcwd
+    temp_dir = tempfile.TemporaryDirectory(prefix=getcwd())
 
-    html_report = join('test_outputs', source, 'report', 'model_comparison_report.html')
+    old_file_dict = {'experiment_dir': 'data/experiments/lr-self-summary-dict/lr-subgroups'}
 
-    output_dir = join('test_outputs', source, 'output')
-    expected_output_dir = join(rsmtool_test_dir, 'data', 'experiments', source, 'output')
+    new_file_dict = copy_data_files(temp_dir.name,
+                                    old_file_dict)
```
Comment (desilinguist) [inline tests/test_experiment_rsmsummarize.py]: Don't need a newline between arguments.

File: tests/test_test_utils.py [hunk 10]
```diff
@@ -24,6 +24,34 @@ def test_copy_data_files(self):
...
+        expected_dict = {'exp_dir': 'temp_test_copy_mixed/lr-subgroups',
+                         'test': 'temp_test_copy_mixed/test.csv'}
+        self.dirs_to_remove.append('temp_test_copy_mixed')
+        output_dict = copy_data_files('temp_test_copy_mixed',
+                                      file_dict)
+        for file_type in expected_dict:
+            eq_(output_dict[file_type], expected_dict[file_type])
+            ok_(Path(output_dict[file_type]).exists())
+        ok_(Path(output_dict['exp_dir']).is_dir())
+        ok_(Path(output_dict['test']).is_file())
```
Comment (desilinguist) [inline tests/test_test_utils.py]: Very nice! Thanks for adding new tests!

File: rsmtool/rsmcompare.py [hunk 11]
```diff
@@ -84,57 +91,72 @@ def run_comparison(config_file_or_obj, output_dir):
...
+        configuration = parser.load_normalize_and_validate_config_from_dict(config_file_or_obj_or_dict,
+                                                                            context='rsmcompare')
+
+    elif isinstance(config_file_or_obj_or_dict, Configuration):
+
+        configuration = config_file_or_obj_or_dict
+        # raise an error if we are passed a Configuration object
+        # without a configdir attribute. This can only
+        # happen if the object was constructed using an earlier version
+        # of RSMTool and stored
```
[the remark this answers is not in the document]
Comment (aloukina) [reply rsmtool/rsmcompare.py]: I meant the package - the object would have to have been constructed using `Configuration` class from API: we don't expose/store it from the command line scripts.

[the remark this answers is not in the document]
Comment (aloukina) [reply rsmtool/rsmcompare.py]: See in other window. This refers to package, not command line script.

File: rsmtool/rsmeval.py [hunk 12]
```diff
@@ -59,25 +67,37 @@ def run_evaluation(config_file_or_obj, output_dir):
...
-            configpath = dirname(configuration.filepath)
-        else:
-            configpath = os.getcwd()
+    elif isinstance(config_file_or_obj_or_dict, Configuration):
+
+        configuration = config_file_or_obj_or_dict
+        # raise an error if we are passed a Configuration object
+        # without a configdir attribute. This can only
+        # happen if the object was constructed using an earlier version
+        # of RSMTool and stored
```
[the remark this answers is not in the document]
Comment (aloukina) [reply rsmtool/rsmeval.py]: same answer

File: rsmtool/rsmpredict.py [hunk 13]
```diff
@@ -57,46 +65,60 @@ def compute_and_save_predictions(config_file_or_obj, output_file, feats_file=Non
...
-        config = config_file_or_obj
-        if config.filepath is not None:
-            configpath = dirname(config.filepath)
-        else:
-            configpath = os.getcwd()
+        configuration = config_file_or_obj_or_dict
+        # raise an error if we are passed a Configuration object
+        # without a configdir attribute. This can only
+        # happen if the object was constructed using an earlier version
+        # of RSMTool and stored
```
[the remark this answers is not in the document]
Comment (aloukina) [reply rsmtool/rsmpredict.py]: same answer

File: rsmtool/rsmsummarize.py [hunk 14]
```diff
@@ -122,26 +131,37 @@ def run_summary(config_file_or_obj, output_dir):
...
-            configpath = dirname(configuration.filepath)
-        else:
-            configpath = os.getcwd()
+    elif isinstance(config_file_or_obj_or_dict, Configuration):
 
+        configuration = config_file_or_obj_or_dict
+        # raise an error if we are passed a Configuration object
+        # without a configdir attribute. This can only
+        # happen if the object was constructed using an earlier version
+        # of RSMTool and stored
```
[the remark this answers is not in the document]
Comment (aloukina) [reply rsmtool/rsmsummarize.py]: same answer
